### PR TITLE
Add relative path support for malioc_diff.

### PR DESCRIPTION
**Pull request by godofredoc**
Relative paths are required for integrating this tool as a test in engine v2.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: impeller/tools/malioc_diff.py [hunk 1]
```diff
@@ -277,6 +283,17 @@ def compare_shaders(malioc_tree, before_shader, after_shader):
 
 def main(argv):
   args = parse_args(argv[1:])
+
+  # Generate full paths if relative ones are provided.
+  args.before = (
+      args.before if os.path.isabs(args.before) else
+      os.path.join(BUILD_ROOT_DIR, args.before)
```
Comment (zanderso) [inline impeller/tools/malioc_diff.py:290]: Someone running the script by hand would expect that relative paths would be relative to their current working directory and not some other directory. Why does this have to be hardcoded to `BUILD_ROOT_DIR` and not the users cwd?

Comment (godofredoc) [reply impeller/tools/malioc_diff.py:290]: The CI builds change the cwd context in multiple places which will make the execution of this script inconsistent depending of where is being run. From the build configuration file perspective we need a consistent way to define relative paths across generators and tests that need to access code, out and third_party dependencies.

Comment (godofredoc) [reply impeller/tools/malioc_diff.py:290]: The build configurations can't resolve to full paths as all the builds run from temporary sandboxes and real paths change all the time. Any specific names that you would like for the additional params?

Comment (godofredoc) [reply impeller/tools/malioc_diff.py:290]: I'm thinking on using `before-relative-to-checkout` and `after-relative-to-checkout`

Comment (godofredoc) [reply impeller/tools/malioc_diff.py]: Done

File: impeller/tools/malioc_diff.py [hunk 2]
```diff
@@ -277,6 +299,16 @@ def compare_shaders(malioc_tree, before_shader, after_shader):
 
 def main(argv):
   args = parse_args(argv[1:])
+
+  # Generate full paths if relative ones are provided with before and
+  # after taking precedence.
+  args.before = (
```
Comment (zanderso) [inline impeller/tools/malioc_diff.py:305]: How does this fail when neither `--before` nor `--before-relative-to-src` is provided. (Same question for the `after` versions.) Maybe putting the fallback logic in the argument validation in `validate_args` would be more clear?

Comment (godofredoc) [reply impeller/tools/malioc_diff.py]: Moved it to the validation method and added some extra validations.

File: impeller/tools/malioc_diff.py [hunk 3]
```diff
@@ -277,6 +316,7 @@ def compare_shaders(malioc_tree, before_shader, after_shader):
 
 def main(argv):
   args = parse_args(argv[1:])
+
```
Comment (zanderso) [inline impeller/tools/malioc_diff.py:319]: nit: remove newline.

**Review by zanderso — APPROVED** [review]
LGTM. Thanks!

File: impeller/tools/malioc_diff.py (hunk 3 above)
Comment (godofredoc) [reply impeller/tools/malioc_diff.py]: done

**Pull request merged**